source/core/timed-out.ts

```diff
@@ -69,7 +71,13 @@ export default function timedOut(request: ClientRequest, delays: Delays, options
 	const {host, hostname} = options;
 
 	const timeoutHandler = (delay: number, event: string): void => {
-		request.destroy(new TimeoutError(delay, event));
+		// Use queueMicroTask to allow for any cancelled events to be handled first,
+		// to prevent firing any TimeoutError unneeded when the event loop is busy or blocked
+		queueMicrotask(() => {
```

Comment:
I don't think `queueMicrotask` will work.

![CleanShot 2025-01-30 at 10 03 36@2x](https://github.com/user-attachments/assets/028582fb-fb9b-4590-87d5-cc91ecd1569c)

[source](https://miro.medium.com/v2/resize:fit:4800/format:webp/1*2yXbhvpf1kj5YT-m_fXgEQ.png)

Ignoring the fact that the above image isn't a perfect representation of what's happening (in Node, other runtimes may do something even more different), what we need to happen is to go around the outer loop. Since we're in the "timer callbacks" phase when we enter this handler, by queueing up another `setTimeout` we're going to do one full outer loop. This means we process I/O events (like connections!) and guarantee whether or not the connection truly was established.

Reply:
Given we've had this patch running in production for a long time, I'm inclined to say it's just the tests that need some tweaking 🙂 

I took a super quick look, and two things stand out to me:
1. `socket timeout` appears to set a custom `setTimeout` on the stream, which may not play nicely with this.
2. `send timeout` ticks a clock. Not sure how that works, but maybe we need two calls to tick to ensure the additional callback is called.

Reply:
I also look into the tests, and the point is that (even though it works) it _does_ break backwards compatibility as before the `TimeoutError` would be thrown synchronously, and now it no longer is returned directly.

We can tune the tests so that is passes, but not sure if that would be acceptable or that we would need a different approach.

Reply:
Ah, fair point :)

Of course @sindresorhus can make the final call, I'd be happy to offer my two cents:

1. The original handler was already part of a `setTimeout`, and hence "async". There were no guarantees that it would trigger the timeout exactly as specified.
2. Doing another `setTimeout` does mean we extend the time for real connection timeouts a bit, since we have to go through the event loop one more time, but we were okay with that tradeoff (because it was sufficiently rare).
3. If backwards compatibility is a must, adding configuration is always an option!

Reply:
Thanks for your input, appreciated.
I think it's up to @sindresorhus to make a fair judgement on this.

Reply:
I don't think anyone should have relied on it being sync. So I'm fine with you just adjusting the tests.

Reply:
Ok, i will look into this 👍 thanks

Reply:
I adjusted the tests so all pass now.

Reply:
The tests are now fixed @sindresorhus @thegedge